### PR TITLE
feat: add test and fixture info to the TestController object and fixture hooks(closes #2826)

### LINKED ISSUE #2826: Allow to use test and fixture names inside hooks and test bodies
### Are you requesting a feature or reporting a bug?
Feature

### What is the current behavior?
You can't access current test and fixture name from the test code.

### What is the expected behavior?
Current test and fixture name can be retrieved by the code running in before/after hooks and the test body function.

### PR DESCRIPTION
**Pull request by Artem-Babich**
## Purpose
#2826 

Currently, you cannot access any test/fixture information inside test and fixture hooks using public API. Fixture hooks cannot access this info even with private API.

## Approach
Add corresponding members fixture: FixtureInfo, test:TestInfo to the TestController instance. Add the fixture info as a second argument to the fixture specific hooks.

## References
#2826 

## API
```ts
type Metadata = Record<string, string>;

interface TestInfo {
    name: string;
    meta: Metadata;
}

interface FixtureInfo {
    name: string;
    meta: Metadata;
    path: string;
}
```

```ts
fixture`testFixture`
    .before(async (ctx, info: FixtureInfo) => {})
    .after(async (ctx, info: FixtureInfo) => {})
    .beforeEach(async t => {
        const testInfo: TestInfo = t.test
        const fixtureInfo: FixtureInfo = t.fixture;
    })
    .afterEach(async t => {
        const testInfo: TestInfo = t.test
        const fixtureInfo: FixtureInfo = t.fixture;
    })

test('SomeTest', async t => {})
    .before(async t => {
        const testInfo: TestInfo = t.test
        const fixtureInfo: FixtureInfo = t.fixture;
    })
    .after(async t => {
        const testInfo: TestInfo = t.test
        const fixtureInfo: FixtureInfo = t.fixture;
    })
```

## Pre-Merge TODO
- [x] Write tests for your proposed changes
- [x] Add typings 
- [ ] Make sure that existing tests do not fail

### CODE REVIEW

File: src/utils/get-testrun-test-and-fixture-info.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import TestRun from '../test-run';
+
+export function getTestRunTestInfo (testRun: TestRun): object {
+    return {
+        name: testRun.test.name,
+        meta: { ...testRun.test.meta },
+    };
+}
+
+export function getTestRunFixtureInfo (testRun: TestRun): object {
```
Comment (miherlosev) [inline src/utils/get-testrun-test-and-fixture-info.ts:10]: Rename to `getFixtureInfo`.

File: src/utils/get-testrun-test-and-fixture-info.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import TestRun from '../test-run';
```
Comment (miherlosev) [inline src/utils/get-testrun-test-and-fixture-info.ts:1]: Rename this file to `get-test-and-fixture-info.ts`.

File: ts-defs-src/test-api/test-controller.d.ts [hunk 3]
```diff
@@ -101,6 +101,17 @@ interface CustomActions {
     [key: string]: (...args: any[]) => TestControllerPromise
 }
 
+interface TestInfo {
+    name: string;
+    meta: Record<string, string>;
```
Comment (miherlosev) [inline ts-defs-src/test-api/test-controller.d.ts:106]: Create a separate type alias for the `meta` property and use it for `FixtureFn`, `TestFn`, `TestInfo`, and `FixtureInfo` aliases.

**Review by miherlosev — APPROVED** [review]

**Comment by Makavelic** [comment]
This breaks a lot of my tests because I use arrays, booleans, etc as data types as part of the meta object. For example:

test.meta({
	testSkip: ["isBrowserstack", "firefox"]

Now it expects all properties to be string values only. Is this expected?

**Comment by Aleksey28** [comment]
Hi @Makavelic,

It is intended behavior. [The documentation](https://testcafe.io/documentation/403436/guides/intermediate-guides/metadata-and-filtering#define-fixture-metadata) states that meta should have a string type. It was a bug before 3.0.0. Could you describe for what purpose you send an array as a value?

**Comment by Makavelic** [comment]
My project has a custom testcafe runner that parses this value to determine whether a test should be skipped or not. An array is useful because I can add several conditions.  I can refactor my runner code to just parse a string and split on ',' or something to get the desired effect.

**Comment by Aleksey28** [comment]
Hi @Makavelic,

Thank you for reporting this issue. We fixed it and published 3.0.0-rc.2 with this fix.

**Pull request merged**